### DISCUSSION TITLE
Python plots agglomerating, history not working?

### DISCUSSION THREAD

**Discussion by marcdotson** [posit-dev/positron#4588] [Q&A] [answered]
I've searched for issues and discussions related to this problem and haven't found anything. Before opening an issue, I wanted to double check that this isn't hidden away in a more technical issue or discussion that I'm not tracking.

I'm working in a Python-only Positron project and have found that the Python plots are agglomerating: Each plot is being *added* onto the previous plot(s) instead of showing up as a separate plot in the history. This problem:

- persists when I use the "Clear all plots" functionality between plots
- occurs for both matplotlib and seaborn plots

I've duplicated the problem following along with Christoph Scheuch's excellent [matplotlib](https://blog.tidy-intelligence.com/posts/ggplot2-vs-matplotlib/) and [seaborn](https://blog.tidy-intelligence.com/posts/ggplot2-vs-seaborn/) blog posts.

**Comment by marcdotson** [comment]
Strange update: I just started producing seaborn.objects plots and the issue is not present. The plots aren't agglomerating and the plot history is working.

**Comment by isabelizimm** [comment, marked as answer]
Thank you for raising this confusing interaction! 🌻 This is related to the [different matplotlib APIs](https://matplotlib.org/stable/users/explain/figure/api_interfaces.html#matplotlib-application-interfaces-apis). The code in the blog post referenced uses the implicit interface, where you operate directly on `plt.XXXXX`, versus the explicit interface where you generate an axis and figure `fig, ax = plt.subplots()`. While using the implicit interface, Positron allows you to iterate on a single plot, which is the behavior you're seeing.

There are a few options to tell `matplotlib` that you are operating on a different figure:
1. add `plt.clf()` between each `plt.show()` call. This will not give you the film strip in Positron, since you’re essentially working on the same plot, but it will clear the plot so you don’t have one mega plot 👾
2. update the code to be in the explicit interface, which would consist of adding `fig, ax = plt.subplots()` to denote the creation of a new plot. Most anywhere you're doing something like `plt.bar(...)` would need to be update to `ax.bar(...)`. You should still be able to use `plt.show()` to show all the figures.

**Reply by marcdotson** [reply]
Thank you!
